gokitmiddlewares/trakingmiddleware/middleware.go

```diff
@@ -0,0 +1,26 @@
+package trackingmiddleware
+
+import (
+	"context"
+
+	"github.com/arquivei/foundationkit/request"
+	"github.com/arquivei/foundationkit/trace"
+	"github.com/go-kit/kit/endpoint"
+)
+
+// New returns a new endpoint middleware that ensures that the context has a request ID and a trace.
+//
+// The request and trace packages both check the context before adding the information in the context so
+// this middleware is safe for usign with middlewares from the trasport layer that may have placed a
+// trace or a request ID in the context.
+//
+// For the trace to work properly, it's expected that the trace.Setup() functions was called previously.
+func New() endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, r interface{}) (interface{}, error) {
+			ctx = request.WithRequestID(ctx)
+			ctx = trace.WithNewTrace(ctx)
```

Comment:
I think this is a problem, because even though I'm using a tracking middleware, I still have to manually set the Trace in every decode step.

Is it possible to have a `tryGetTraceFromRequest()` function? if the request follows a default structure or interface, we can get tracking for free just by using the "default" way of building microservices.

Reply:
If it is to get the Trace ID from a request struct, it makes sense that each decoder may need to implement it's own, unless we define a convention to how the request struct should be to all endpoints "ever". The middlewares combo was meant to isolate trace recovery into the transport layer so that endpoints could obtain them from the request context. Loading from request body should probably be another type of tracking middleware.

Reply:
I believe the correct way to propagate trace and request ID to the service layer and bellow is by context. I think it's the responsibility of the transport layer to put the trace in the context if it's being received, like in the header of an HTTP request or a body of a rpc call or thought kafka message body. 
This middleware is to ensure the info is in the context in the endpoint layer, because there might not be a transport middleware in some cenario.